File: src/main/java/leets/weeth/domain/user/domain/repository/UserRepository.java
package leets.weeth.domain.user.domain.repository;

import leets.weeth.domain.user.domain.entity.Cardinal;
import leets.weeth.domain.user.domain.entity.User;
import leets.weeth.domain.user.domain.entity.enums.Status;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface UserRepository extends JpaRepository<User, Long> {

    Optional<User> findByEmail(String email);

    Optional<User> findByKakaoId(long kakaoId);

    Optional<User> findByAppleId(String appleId);

    List<User>findAllByNameContainingAndStatus(String name, Status status);

    boolean existsByEmail(String email);

    boolean existsByStudentId(String studentId);

    boolean existsByTel(String tel);

    boolean existsByStudentIdAndIdIsNot(String studentId, Long id);

    boolean existsByTelAndIdIsNot(String tel, Long id);

    List<User> findAllByStatusOrderByName(Status status);

    List<User> findAllByOrderByNameAsc();

    @Query("SELECT uc.user FROM UserCardinal uc WHERE uc.cardinal = :cardinal AND uc.user.status = :status")
    List<User> findAllByCardinalAndStatus(@Param("cardinal") Cardinal cardinal, @Param("status") Status status);

    /*
    todo 차후 리팩토링
     */
    @Query("""
                SELECT u
                FROM User u
                JOIN UserCardinal uc ON u.id = uc.user.id
                JOIN uc.cardinal c
                WHERE u.status = :status
                GROUP BY u.id
                ORDER BY MAX(c.cardinalNumber) DESC, u.name ASC
            """)
    Slice<User> findAllByStatusOrderedByCardinalAndName(@Param("status") Status status, Pageable pageable);

    @Query("""
                SELECT u FROM User u
                JOIN UserCardinal uc ON uc.user.id = u.id
                WHERE u.status = :status
                AND uc.cardinal = :cardinal
                ORDER BY u.name ASC
            """)
    Slice<User> findAllByCardinalOrderByNameAsc(@Param("status") Status status, @Param("cardinal") Cardinal cardinal, Pageable pageable);
}

File: src/main/kotlin/com/weeth/domain/user/domain/repository/UserRepository.kt
package com.weeth.domain.user.domain.repository

import com.weeth.domain.user.application.exception.UserNotFoundException
import com.weeth.domain.user.domain.entity.User
import com.weeth.domain.user.domain.vo.Email
import com.weeth.global.common.vo.PhoneNumber
import jakarta.persistence.LockModeType
import jakarta.persistence.QueryHint
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import org.springframework.data.jpa.repository.QueryHints
import org.springframework.data.repository.query.Param
import java.util.Optional

interface UserRepository :
    JpaRepository<User, Long>,
    UserReader {
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @QueryHints(QueryHint(name = "jakarta.persistence.lock.timeout", value = "2000"))
    @Query("SELECT u FROM User u WHERE u.id = :id")
    fun findByIdWithLock(
        @Param("id") id: Long,
    ): Optional<User>

    fun findByEmail(email: Email): Optional<User>

    fun existsByEmail(email: Email): Boolean

    fun existsByStudentId(studentId: String): Boolean

    fun existsByTel(tel: PhoneNumber): Boolean

    fun existsByStudentIdAndIdIsNot(
        studentId: String,
        id: Long,
    ): Boolean

    fun existsByTelAndIdIsNot(
        tel: PhoneNumber,
        id: Long,
    ): Boolean

    fun findAllByOrderByNameAsc(): List<User>

    fun findByEmailValue(email: String): Optional<User> = findByEmail(Email.from(email))

    fun existsByEmailValue(email: String): Boolean = existsByEmail(Email.from(email))

    fun existsByTelValue(tel: String): Boolean = existsByTel(PhoneNumber.from(tel))

    fun existsByTelAndIdIsNotValue(
        tel: String,
        id: Long,
    ): Boolean = existsByTelAndIdIsNot(PhoneNumber.from(tel), id)

    override fun getById(userId: Long): User = findById(userId).orElseThrow { UserNotFoundException() }

    override fun getByIdWithLock(userId: Long): User = findByIdWithLock(userId).orElseThrow { UserNotFoundException() }

    override fun getByEmail(email: String): User = findByEmailValue(email).orElseThrow { UserNotFoundException() }

    override fun findByIdOrNull(userId: Long): User? = findById(userId).orElse(null)

    override fun findAllByIds(userIds: List<Long>): List<User> = findAllById(userIds)
}
